ko/build/gobuild_test.go

```diff
@@ -0,0 +1,226 @@
+				t.Errorf("computeImportpath() = %v, want error", ip)
+			} else if err == nil && !test.expectErr {
+				if got, want := ip, test.expectImportpath; want != got {
+					t.Errorf("computeImportpath() = %v, want %v", got, want)
+				}
+			}
+		})
+	}
+}
+
+func TestGoBuildIsSupportedRef(t *testing.T) {
+	img, err := random.Image(1024, 1)
+	if err != nil {
+		t.Fatalf("random.Image() = %v", err)
+	}
+	importpath := "github.com/google/go-containerregistry"
+	tc := testContext{
+		gopath:  "/go",
+		workdir: "/go/src/" + importpath,
+	}
+	tc.Enter(t)
+	defer tc.Exit(t)
+	ng, err := NewGo(Options{Base: img})
+	if err != nil {
+		t.Fatalf("NewGo() = %v", err)
+	}
+
+	supportedTests := []string{
+		path.Join(importpath, "pkg", "foo"),
+		path.Join(importpath, "cmd", "d8s"),
+	}
+
+	for _, test := range supportedTests {
+		t.Run(test, func(t *testing.T) {
+			if !ng.IsSupportedReference(test) {
+				t.Errorf("IsSupportedReference(%v) = false, want true", test)
+			}
+		})
+	}
+
+	unsupportedTests := []string{
+		"simple string",
+		"k8s.io/client-go/pkg/foo",
+		"github.com/google/secret/cmd/sauce",
+		path.Join("vendor", importpath, "pkg", "foo"),
+	}
+
+	for _, test := range unsupportedTests {
+		t.Run(test, func(t *testing.T) {
+			if ng.IsSupportedReference(test) {
+				t.Errorf("IsSupportedReference(%v) = true, want false", test)
+			}
+		})
+	}
+}
+
+// A helper method we use to substitute for the default "build" method.
+func writeTempFile(s string) (string, error) {
+	file, err := ioutil.TempFile(os.TempDir(), "out")
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+	if _, err := file.WriteString(s); err != nil {
+		return "", err
+	}
+	return file.Name(), nil
+}
+
+func TestGoBuild(t *testing.T) {
+	baseLayers := int64(3)
+	base, err := random.Image(1024, baseLayers)
+	if err != nil {
+		t.Fatalf("random.Image() = %v", err)
+	}
+	importpath := "github.com/google/go-containerregistry"
+	tc := testContext{
+		gopath:  "/go",
+		workdir: "/go/src/" + importpath,
+	}
+	tc.Enter(t)
+	defer tc.Exit(t)
+	ng, err := NewGo(Options{Base: base})
+	if err != nil {
+		t.Fatalf("NewGo() = %v", err)
+	}
+	ng.(*gobuild).build = writeTempFile
+
+	img, err := ng.Build(path.Join(importpath, "cmd", "d8s"))
+	if err != nil {
+		t.Errorf("Build() = %v", err)
+	}
+
+	ls, err := img.Layers()
+	if err != nil {
+		t.Errorf("Layers() = %v", err)
+	}
+
+	// Check that we have the expected number of layers.
+	t.Run("check layer count", func(t *testing.T) {
+		if got, want := int64(len(ls)), baseLayers+1; got != want {
+			t.Fatalf("len(Layers()) = %v, want %v", got, want)
+		}
+	})
+
+	// While we have a randomized base image, the application layer should be completely deterministic.
+	// Check that when given fixed build outputs we get a fixed layer hash.
+	t.Run("check determinism", func(t *testing.T) {
+		expectedHash := v1.Hash{
+			Algorithm: "sha256",
+			Hex:       "f6c290fa444e8e94936132309a42764f8360fb88e38adb8b3077a21780128794",
```

Comment:
Where does this magic value come from? Can you derive it in the test itself? Or to make the point about determinism, build the thing twice and show that the digest matches, whatever it happens to be.